### PR TITLE
support customer-supplied encryption keys

### PR DESCRIPTION
**Pull request by xxchan**
part of aos-dev/go-storage#523

ref: https://cloud.google.com/storage/docs/encryption

- Google-managed encryption keys: default behaviour, no setup or configuration required
- customer-supplied encryption keys: has the following HTTP headers:
  - x-goog-encryption-algorithm
  - x-goog-encryption-key
  - x-goog-encryption-sha256
  - copy operation also has encryption headers, but we havn't supported `copy` for gcs yet.
- Customer-managed encryption keys: service agent performs encryption and decryption with customer-managed encryption keys

### CODE REVIEW

File: storage.go [hunk 1]
```diff
@@ -138,6 +138,9 @@ func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairSt
 	var rc io.ReadCloser
 
 	object := s.bucket.Object(rp)
+	if opt.HasSseCustomerKey {
+		object = object.Key([]byte(opt.SseCustomerKey))
```
Comment (Xuanwo) [inline storage.go]: Maybe we should let the user pass a `[]byte` directly? Or input a string after `base64`?

File: service.toml [hunk 2]
```diff
@@ -32,4 +36,7 @@ type = "string"
 type = "string"
 
 [infos.object.meta.storage-class]
-type = "string"
\ No newline at end of file
+type = "string"
+
+[infos.object.meta.sse_customer_key_sha256]
+type = "string"
```
Comment (Xuanwo) [inline service.toml]: What's the content of `sse_customer_key_sha256`, a base64 string?

Comment (xxchan) [reply service.toml]: base64-encoded SHA-256 hash of the customer-supplied encryption key for the object

File: storage.go (hunk 1 above)
Comment (xxchan) [reply storage.go]: I agree, but how to support `[]byte` in the toml

Comment (Xuanwo) [reply storage.go]: Do we need to set `CustomerKeySHA256` before we write data into it?

Comment (Xuanwo) [reply storage.go]: > I agree, but how to support []byte in the toml

Well, it looks like not supported for now.

We can register a new type here: https://github.com/aos-dev/go-storage/blob/master/cmd/definitions/parse.go#L77

Maybe `byte_array`?

Comment (Xuanwo) [reply storage.go]:
```go
func setEncryptionHeaders(headers http.Header, key []byte, copySource bool) error {
	if key == nil {
		return nil
	}
	// TODO(jbd): Ask the API team to return a more user-friendly error
	// and avoid doing this check at the client level.
	if len(key) != 32 {
		return errors.New("storage: not a 32-byte AES-256 key")
	}
	var cs string
	if copySource {
		cs = "copy-source-"
	}
	headers.Set("x-goog-"+cs+"encryption-algorithm", "AES256")
	headers.Set("x-goog-"+cs+"encryption-key", base64.StdEncoding.EncodeToString(key))
	keyHash := sha256.Sum256(key)
	headers.Set("x-goog-"+cs+"encryption-key-sha256", base64.StdEncoding.EncodeToString(keyHash[:]))
	return nil
}
```

It looks like gcs's sdk will set it.

File: utils.go [hunk 3]
```diff
@@ -277,6 +277,9 @@ func (s *Storage) formatFileObject(v *gs.ObjectAttrs) (o *typ.Object, err error)
 	if value := v.StorageClass; value != "" {
 		sm[MetadataStorageClass] = value
 	}
+	if value := v.CustomerKeySHA256; value != "" {
+		sm[MetadataSseCustomerKeySha256] = value
```
Comment (Xuanwo) [inline utils.go]: How about keeping the naming style with SDK?

- `sse_customer_key` => `customer_key`
- `sse_customer_key_sha256` => `customer_key_sha256`

Or the naming in GCS's docs:

- `encryption_key`
- `encryption_key_sha256`

Comment (Xuanwo) [reply utils.go]: I wrote a RFC about this: https://github.com/aos-dev/specs/pull/38

**Comment by xxchan** [comment]
I haven't completely understand how to use gcs' KMS. It seems we don't need to do anything here?

**Comment by Xuanwo** [comment]
> I haven't completely understand how to use gcs' KMS. It seems we don't need to do anything here?

Yes.

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**